File: FWFramework/Classes/Component/Tracking/FWAuthorizeTracking.m
//
//  FWAuthorize.m
//  FWFramework
//
//  Created by wuyong on 17/3/15.
//  Copyright © 2018年 wuyong.site. All rights reserved.
//

#import "FWAuthorizeTracking.h"
#import <AdSupport/ASIdentifierManager.h>
#if __IPHONE_14_0
#import <AppTrackingTransparency/AppTrackingTransparency.h>
#endif
#if FWFrameworkSPM
@import FWFramework;
#else
#import "FWAuthorize.h"
#endif

// iOS14+使用AppTrackingTransparency，其它使用AdSupport
@interface FWAuthorizeTracking : NSObject <FWAuthorizeProtocol>

@end

@implementation FWAuthorizeTracking

+ (void)load
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        [FWAuthorizeManager registerAuthorize:FWAuthorizeTypeTracking withBlock:^id<FWAuthorizeProtocol>{
            return [[FWAuthorizeTracking alloc] init];
        }];
    });
}

- (FWAuthorizeStatus)authorizeStatus
{
#if __IPHONE_14_0
    if (@available(iOS 14.0, *)) {
        ATTrackingManagerAuthorizationStatus status = ATTrackingManager.trackingAuthorizationStatus;
        switch (status) {
            case ATTrackingManagerAuthorizationStatusRestricted:
                return FWAuthorizeStatusRestricted;
            case ATTrackingManagerAuthorizationStatusDenied:
                return FWAuthorizeStatusDenied;
            case ATTrackingManagerAuthorizationStatusAuthorized:
                return FWAuthorizeStatusAuthorized;
            case ATTrackingManagerAuthorizationStatusNotDetermined:
            default:
                return FWAuthorizeStatusNotDetermined;
        }
    }
#endif
    
    return [ASIdentifierManager sharedManager].isAdvertisingTrackingEnabled ? FWAuthorizeStatusAuthorized : FWAuthorizeStatusDenied;
}

- (void)authorize:(void (^)(FWAuthorizeStatus status))completion
{
#if __IPHONE_14_0
    if (@available(iOS 14.0, *)) {
        [ATTrackingManager requestTrackingAuthorizationWithCompletionHandler:^(ATTrackingManagerAuthorizationStatus status) {
            if (completion) {
                dispatch_async(dispatch_get_main_queue(), ^{
                    completion(self.authorizeStatus);
                });
            }
        }];
        return;
    }
#endif
    
    if (completion) {
        dispatch_async(dispatch_get_main_queue(), ^{
            completion(self.authorizeStatus);
        });
    }
}

@end

File: FWFramework/Classes/Module/Tracking/FWAuthorizeTracking.swift
//
//  FWAuthorizeTracking.swift
//  FWFramework
//
//  Created by wuyong on 2019/6/27.
//  Copyright © 2019 wuyong.site. All rights reserved.
//

import UIKit
import AdSupport
import AppTrackingTransparency
#if FWFrameworkSPM
import FWFramework
import FWFrameworkCompatible
#endif

@objc extension UIDevice {
    /// 获取设备IDFA(外部使用)，重置广告或系统后会改变，需先检测广告追踪权限，启用Tracking子模块后生效
    public class var fwDeviceIDFA: String {
        return ASIdentifierManager.shared().advertisingIdentifier.uuidString
    }
}

/// IDFA授权，iOS14+使用AppTrackingTransparency，其它使用AdSupport
private class FWAuthorizeTracking: NSObject, FWAuthorizeProtocol {
    func authorizeStatus() -> FWAuthorizeStatus {
        if #available(iOS 14.0, *) {
            let status = ATTrackingManager.trackingAuthorizationStatus
            switch status {
            case .restricted:
                return .restricted
            case .denied:
                return .denied
            case .authorized:
                return .authorized
            default:
                return .notDetermined
            }
        } else {
            return ASIdentifierManager.shared().isAdvertisingTrackingEnabled ? .authorized : .denied
        }
    }
    
    func authorize(_ completion: ((FWAuthorizeStatus) -> Void)?) {
        if #available(iOS 14.0, *) {
            ATTrackingManager.requestTrackingAuthorization { status in
                if completion != nil {
                    DispatchQueue.main.async { [weak self] in
                        guard let self = self else { return }
                        completion?(self.authorizeStatus())
                    }
                }
            }
        } else {
            if completion != nil {
                DispatchQueue.main.async { [weak self] in
                    guard let self = self else { return }
                    completion?(self.authorizeStatus())
                }
            }
        }
    }
}

@objc extension FWAutoloader {
    private func loadAuthorizeTracking() {
        FWAuthorizeManager.registerAuthorize(.tracking) {
            return FWAuthorizeTracking()
        }
    }
}
